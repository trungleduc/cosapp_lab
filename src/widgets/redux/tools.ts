import { getNodeAtPath, walk } from 'react-sortable-tree';

/**
 *
 * Helper function to call `fn` each `ms` millisecond
 * @export
 * @param {*} fn
 * @param {*} ms
 * @param {*} number
 * @returns {*}
 */

export function debounce(fn: any, ms: number): any {
  let timer;
  return () => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      fn();
    }, ms);
  };
}

/**
 *
 *
 * @export
 * @param {*} treeData
 * @param {number[]} path
 * @returns
 */
export function getNodeFullPath(treeData: any, path: number[]) {
  const getNodeKey = (obj: { treeIndex: any }) => obj.treeIndex;
  let fullPath: string;
  for (let index = 0; index < path.length; index++) {
    const newPath = path.slice(0, index + 1);
    const currentNode = getNodeAtPath({
      treeData: treeData,
      path: newPath,
      getNodeKey: getNodeKey
    });
    const currentTitle = currentNode.node.title;
    if (index == 0) {
      fullPath = currentTitle as string;
    } else {
      fullPath = fullPath.concat('.' + currentTitle);
    }
  }
  return fullPath;
}

export function updateNodeId(treeData: any, oldId: string, newId: string) {
  const getNodeKey = (obj: { treeIndex: any }) => obj.treeIndex;
  const ret = {};
  const callback = item => {
    const nodeId: string = item.node.id;
    if (nodeId.startsWith(oldId)) {
      const newNodeId = nodeId.replace(oldId, newId);
      item.node.id = newNodeId;
      ret[nodeId] = newNodeId;
    }
  };
  walk({ treeData, getNodeKey, callback });

  return ret;
}
