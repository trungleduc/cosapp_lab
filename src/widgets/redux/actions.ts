import {
  Action,
  ActionType,
  StateInterface,
  ArchAddNode,
  MainSwitchPanel,
  PanelStatusInterface,
  ArchSaveGraphJson,
  ArchSaveGraphPosition,
  ArchFilterNode,
  ArchRemoveNode,
  ArchSelectNode,
  ArchUpdateTree,
  ArchUpdateConnectionGraph,
  UserListType,
  ServerGetData,
  DashboardAddData,
  DashboardSetController,
  DashboardAddController,
  DashboardRemoveController,
  DashboardUpdateController,
  DashboardUpdateComputedResult
} from './types';
import { addNodeUnderParent, removeNodeAtPath } from 'react-sortable-tree';
import { getNodeFullPath } from './tools';
/**
 * Reset store to initial state
 */
export function resetStore(): ActionType {
  return { type: Action.RESET_STORE };
}

/**
 *
 *
 * Send save state signal to all component
 * @returns {ActionType}
 */
export function saveState(): ActionType {
  return { type: Action.SAVE_STATE };
}

/**
 *
 *
 * Action to update saveSignal of state
 * @param {StateInterface} state
 */
export function saveState_(state: StateInterface) {
  const oldSignal = state.saveSignal;
  const newState: StateInterface = { ...state, saveSignal: oldSignal + 1 };
  return newState;
}

/**
 * Action to switch activated panel
 * @param panelName - Name of requested panel
 */
export function mainSwitchPanel(panelName: string): ActionType {
  return { type: Action.SWITCH_PANEL, panelName };
}

/**
 * Undate store for panel activation. This function is called
 * when `mainSwitchPanel` is dispatched
 * @param state
 * @param action
 */
export function mainSwitchPanel_(
  state: StateInterface,
  action: MainSwitchPanel
): StateInterface {
  const newPanelStatus: PanelStatusInterface = {
    panel_0: [false, 'Project_Overview'],
    panel_1: [false, 'System_Arc'],
    panel_2: [false, 'Module_Creator'],
    panel_3: [false, 'Dashboard'],
    panel_4: [false, 'User_Management'],
    panel_5: [false, 'Module_Library']
  };
  for (const key in newPanelStatus) {
    if (newPanelStatus.hasOwnProperty(key)) {
      newPanelStatus[key][0] = action.panelName === newPanelStatus[key][1];
    }
  }

  return {
    ...state,
    mainState: newPanelStatus
  };
}

/**
 * Lock/ unlock diagram.
 * @param status
 */
export function archLockGraph(status: boolean): ActionType {
  return { type: Action.LOCK_GRAPH, status };
}

/**
 * Action to save data from backend to redux store.
 * @param data
 */
export function serverGetData(data: any): ActionType {
  return { type: Action.GET_SERVER_DATA, data };
}

/**
 * Update store with data from backend. This function is
 * called when `serverGetData` is dispatched
 * @param state
 * @param action
 */
export function serverGetData_(state: StateInterface, action: ServerGetData) {
  return {
    ...state,
    systemArch: {
      ...state.systemArch,
      systemGraph: action.data.systemGraph,
      systemTree: {
        ...state.systemArch.systemTree,
        nodeData: action.data.systemTree
      }
    },
    dashboardState: {
      ...state.dashboardState,
      variableData: action.data.variableData
    }
  };
}

/**
 *
 * Save diagram json data to state
 * @export
 * @param {{[key:string]: any}} data
 * @returns {ActionType}
 */
export function archSaveGraphJson(data: { [key: string]: any }): ActionType {
  return { type: Action.ARCH_SAVE_GRAPH_JSON, data };
}

/**
 *
 * This function is called when archSaveGraphJson is dispatched
 * @export
 * @param {StateInterface} state
 * @param {ArchSaveGraphJson} action
 * @returns {StateInterface}
 */
export function archSaveGraphJson_(
  state: StateInterface,
  action: ArchSaveGraphJson
): StateInterface {
  const connectionData = action.data.connection;
  const positionData = action.data.position;
  const new_state = { ...state };

  for (const sysKey in new_state.systemArch.systemGraph.systemGraphData) {
    if (connectionData.hasOwnProperty(sysKey)) {
      new_state.systemArch.systemGraph.systemGraphData[sysKey]['connections'] =
        connectionData[sysKey];
    }
    if (positionData.hasOwnProperty(sysKey)) {
      new_state.systemArch.systemGraph.systemGraphData[sysKey]['position'] =
        positionData[sysKey];
    }
  }
  const stringData = JSON.stringify(new_state, null, 2);

  const blob = new Blob([stringData], {
    type: 'application/json'
  });
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = 'store.json';

  a.click();
  return { ...state };
}

/**
 *
 * Save the position of all node in pbs graph to store
 * @export
 * @param {{
 *   [key: string]: Array<number>;
 * }} data
 * @returns {ActionType}
 */
export function archSaveGraphPosition(data: {
  [key: string]: { visible: boolean; position: Array<number> };
}): ActionType {
  return { type: Action.ARCH_SAVE_GRAPH_POSITION, data };
}

/**
 *
 * This function is called when archSaveGraphPosition is dispatched
 * @export
 * @param {StateInterface} state
 * @param {ArchSaveGraphPosition} action
 * @returns {StateInterface}
 */
export function archSaveGraphPosition_(
  state: StateInterface,
  action: ArchSaveGraphPosition
): StateInterface {
  return {
    ...state,
    systemArch: { ...state.systemArch, systemPBS: action.data }
  };
}

/**
 *
 * Modify the visibility flag of node data in store
 * @export
 * @param {Array<string>} data
 * @returns {ActionType}
 */
export function archFilterNode(
  data: Array<string>,
  selected: Array<string>
): ActionType {
  return { type: Action.ARCH_FILTER_NODE, data, selected };
}

/**
 *
 * This function is called when archFilterNode is dispatched
 * @export
 * @param {StateInterface} state
 * @param {ArchFilterNode} action
 * @returns {StateInterface}
 */
export function archFilterNode_(
  state: StateInterface,
  action: ArchFilterNode
): StateInterface {
  let currentData: {
    [key: string]: { visible: boolean; position: Array<number> };
  };
  currentData = JSON.parse(JSON.stringify(state.systemArch.systemPBS));
  const currentSignal = state.systemArch.systemPBSUpdated;
  const selected = action.selected;
  if (action.data.length > 0) {
    for (const key in currentData) {
      currentData[key].visible = false;
    }
    action.data.forEach(nodeName => {
      currentData[nodeName].visible = true;
    });
  } else {
    for (const key in currentData) {
      currentData[key].visible = true;
    }
  }
  return {
    ...state,
    systemArch: {
      ...state.systemArch,
      systemTree: { ...state.systemArch.systemTree, nodePath: selected },
      systemPBS: currentData,
      systemPBSUpdated: currentSignal + 1
    }
  };
}

/**
 * Update connection graph data from tree data
 *
 * @export
 * @param {string} oldId
 * @param {string} newId
 * @returns {ActionType}
 */
export function archUpdateConnectionGraph(
  data: { [key: string]: string },
  root: string
): ActionType {
  return { type: Action.ARCH_UPDATE_CONNECTION_GRAPH, data, root };
}

/**
 * This function is called when `archUpdateConnectionGraph` is dispatched
 * @param state
 * @param action
 */
export function archUpdateConnectionGraph_(
  state: StateInterface,
  action: ArchUpdateConnectionGraph
): StateInterface {
  const updateData = action.data;

  const root = action.root;
  const { systemGraphData, systemList } = { ...state.systemArch.systemGraph };
  const newSystemGraphData = { ...systemGraphData };
  let newSystemList = [...systemList];

  const rootPath = root.split('.');
  if (rootPath.length > 1) {
    const rootChild = rootPath[rootPath.length - 1];
    const rootParent = root.replace(`.${rootChild}`, '');

    const newConnection = newSystemGraphData[rootParent].connections.filter(
      (value: string[]) =>
        !(value[0].includes(rootChild) || value[1].includes(rootChild))
    );
    newSystemGraphData[rootParent].connections = newConnection;
  }

  for (const [oldId, newId] of Object.entries(updateData)) {
    if (oldId in systemGraphData) {
      newSystemGraphData[newId] = systemGraphData[oldId];
      delete newSystemGraphData[oldId];
    }
  }
  newSystemList = Object.keys(newSystemGraphData);
  const newNodePath = [];
  state.systemArch.systemTree.nodePath.forEach(element => {
    if (Object.keys(updateData).includes(element)) {
      newNodePath.push(updateData[element]);
    }
  });

  return {
    ...state,
    systemArch: {
      ...state.systemArch,
      systemTree: { ...state.systemArch.systemTree, nodePath: newNodePath },
      systemGraph: {
        ...state.systemArch.systemGraph,
        systemGraphData: newSystemGraphData,
        systemList: newSystemList,
        updateData
      }
    }
  };
}

/**
 *
 *
 * @export
 * @param {string} nodeId
 * @returns {ActionType}
 */
export function archSelectNode(nodeId: string): ActionType {
  return { type: Action.SELECT_NODE, nodeId };
}

/**
 *
 *
 * @export
 * @param {StateInterface} state
 * @param {ArchSelectNode} action
 * @returns {StateInterface}
 */
export function archSelectNode_(
  state: StateInterface,
  action: ArchSelectNode
): StateInterface {
  return {
    ...state,
    systemArch: {
      ...state.systemArch,
      systemTree: {
        ...state.systemArch.systemTree,
        selectedNode: action.nodeId
      }
    }
  };
}

/**
 * Add a new node to system structure tree
 * @param nodeName
 * @param nodePath
 * @param nodeData
 */
export function archAddNode(
  nodeName: string,
  nodePath: string,
  nodeData: {
    personName: UserListType[];
    sysClass: string;
    portList: {
      portName: string;
      portDirection: number;
      portClass: string;
    }[];
  }
): ActionType {
  return { type: Action.ADD_NODE, nodeName, nodePath, nodeData };
}

/**
 * This function is called when `archAddNode` is dispatched
 * @param state
 * @param action
 */
export function archAddNode_(
  state: StateInterface,
  action: ArchAddNode
): StateInterface {
  const getNodeKey = (obj: { treeIndex: any }) => obj.treeIndex;

  const path = action.nodePath;
  const fullPath = getNodeFullPath(state.systemArch.systemTree.nodeData, path);
  const newTreeData = addNodeUnderParent({
    treeData: state.systemArch.systemTree.nodeData,
    parentKey: path[path.length - 1],
    expandParent: true,
    getNodeKey,
    newNode: {
      title: action.nodeName
    }
  }).treeData;
  const newNodeName = action.nodeName;
  const fullNodeName = fullPath.concat('.' + newNodeName);

  const inPortList: Array<string> = ['inwards'];
  const outPortList: Array<string> = ['outwards'];
  const connectionList: Array<Array<string>> = [];
  for (let index = 0; index < action.nodeData.portList.length; index++) {
    const element = action.nodeData.portList[index];
    if (element.portDirection === 0) {
      inPortList.push(element.portName);
    } else {
      outPortList.push(element.portName);
    }
  }
  return {
    ...state,
    systemArch: {
      ...state.systemArch,
      systemTree: {
        ...state.systemArch.systemTree,
        nodeData: newTreeData
      },
      systemGraph: {
        ...state.systemArch.systemGraph,
        systemGraphData: {
          ...state.systemArch.systemGraph.systemGraphData,
          [fullNodeName]: {
            inPort: inPortList,
            outPort: outPortList,
            connections: connectionList
          }
        },
        systemList: [...state.systemArch.systemGraph.systemList, fullNodeName]
      },
      systemData: {
        ...state.systemArch.systemData,
        mainData: {
          ...state.systemArch.systemData.mainData,
          [fullNodeName]: {
            user: action.nodeData.personName,
            class: action.nodeData.sysClass,
            portList: inPortList.concat(outPortList)
          }
        },
        portData: {
          ...state.systemArch.systemData.portData
        }
      }
    }
  };
}

/**
 * Remove the selected node from system structure tree
 * @param nodeData
 * @param nodePath
 */
export function archRemoveNode(nodeData: any, nodePath: any): ActionType {
  return { type: Action.REMOVE_NODE, nodeData, nodePath };
}

/**
 * This function is called when `archRemoveNode` is dispatched
 * @param state
 * @param action
 */
export function archRemoveNode_(
  state: StateInterface,
  action: ArchRemoveNode
): StateInterface {
  const getNodeKey = (obj: { treeIndex: any }) => obj.treeIndex;
  const path = action.nodePath;
  const fullNodeName = getNodeFullPath(
    state.systemArch.systemTree.nodeData,
    path
  );

  const newSystemList = [];
  const newSystemGraphData = {};
  const newSystemData = {};

  const newTreeData = removeNodeAtPath({
    treeData: state.systemArch.systemTree.nodeData,
    path,
    getNodeKey
  });
  state.systemArch.systemGraph.systemList.forEach(fullSysName => {
    if (fullSysName.startsWith(fullNodeName)) {
    } else {
      newSystemList.push(fullSysName);
      if (
        state.systemArch.systemGraph.systemGraphData.hasOwnProperty(fullSysName)
      ) {
        newSystemGraphData[fullSysName] =
          state.systemArch.systemGraph.systemGraphData[fullSysName];
        newSystemData[fullSysName] =
          state.systemArch.systemData.mainData[fullSysName];
      }
    }
  });

  return {
    ...state,
    systemArch: {
      ...state.systemArch,
      systemTree: {
        ...state.systemArch.systemTree,
        nodeData: newTreeData
      },
      systemGraph: {
        ...state.systemArch.systemGraph,
        systemList: newSystemList,
        systemGraphData: newSystemGraphData
      },
      systemData: {
        ...state.systemArch.systemData,
        mainData: newSystemData,
        portData: {
          ...state.systemArch.systemData.portData
        }
      }
    }
  };
}

/**
 * Update structure tree data
 * @param treeData
 */
export function archUpdateTreeData(
  treeData: any,
  updatePbs = false
): ActionType {
  return { type: Action.UPDATE_TREE, treeData, updatePbs };
}

export function archUpdateTreeData_(
  state: StateInterface,
  action: ArchUpdateTree
): StateInterface {
  let pbsUpdateSignal = state.systemArch.systemPBSUpdated;
  if (action.updatePbs) {
    pbsUpdateSignal += 1;
  }

  return {
    ...state,
    systemArch: {
      ...state.systemArch,
      systemTree: {
        ...state.systemArch.systemTree,
        nodeData: action.treeData
      },
      systemPBSUpdated: pbsUpdateSignal
    }
  };
}

/**
 * Update time step list in redux store
 * @param data
 */
export function dashboardAdd3DData(data: any): ActionType {
  return { type: Action.DASHBOARD_ADD_3D_DATA, data };
}

/**
 * This function is called if `dashboardAdd3DData` is dispatched
 * @param state
 * @param action
 */
export function dashboardAdd3DData_(
  state: StateInterface,
  action: DashboardAddData
): StateInterface {
  const newTimeList = ['None', ...action.data];

  return {
    ...state,
    dashboardState: { ...state.dashboardState, timeStepList: newTimeList }
  };
}

/**
 * Toggle run status of store
 */
export function dashboardToggleRun() {
  return { type: Action.DASHBOARD_TOGGLE_RUN };
}

export function dashboardSetController(data: {
  [key: string]: number;
}): ActionType {
  return { type: Action.DASHBOARD_SET_CONTROLLER, data };
}

export function dashboardSetController_(
  state: StateInterface,
  action: DashboardSetController
): StateInterface {
  return {
    ...state,
    dashboardState: { ...state.dashboardState, selectedVariable: action.data }
  };
}

/**
 * Add a parameter controller to store
 * @param variableName
 * @param value
 */
export function dashboardAddController(
  variableName: string,
  value: number
): ActionType {
  return { type: Action.DASHBOARD_ADD_CONTROLLER, variableName, value };
}

/**
 * This function is called if `dashboardAddController` is dispatched
 * @param state
 * @param action
 */
export function dashboardAddController_(
  state: StateInterface,
  action: DashboardAddController
): StateInterface {
  const currentVarlist = Object.keys(state.dashboardState.selectedVariable);
  if (!currentVarlist.includes(action.variableName)) {
    const newVarList = {
      ...state.dashboardState.selectedVariable,
      [action.variableName]: action.value
    };
    return {
      ...state,
      dashboardState: { ...state.dashboardState, selectedVariable: newVarList }
    };
  } else {
    return state;
  }
}

/**
 * Update the value of selected controller
 * @param variableName
 * @param value
 */
export function dashboardUpdateController(
  variableName: string,
  value: number
): ActionType {
  return {
    type: Action.DASHBOARD_UPDATE_CONTROLLER,
    variableName,
    value
  };
}

/**
 * This function is called if `dashboardUpdateController` is dispatched
 * @param state
 * @param action
 */
export function dashboardUpdateController_(
  state: StateInterface,
  action: DashboardUpdateController
): StateInterface {
  if (
    state.dashboardState.selectedVariable.hasOwnProperty(action.variableName)
  ) {
    return {
      ...state,
      dashboardState: {
        ...state.dashboardState,
        selectedVariable: {
          ...state.dashboardState.selectedVariable,
          [action.variableName]: action.value
        }
      }
    };
  } else {
    return state;
  }
}

/**
 * Remove selected controller
 * @param variableName
 */
export function dashboardRemoveController(variableName: string): ActionType {
  return {
    type: Action.DASHBOARD_REMOVE_CONTROLLER,
    variableName: variableName
  };
}

/**
 * This function is called if `dashboardRemoveController` is dispatched
 * @param state
 * @param action
 */
export function dashboardRemoveController_(
  state: StateInterface,
  action: DashboardRemoveController
): StateInterface {
  const {
    [action.variableName]: value,
    ...newVarList
  } = state.dashboardState.selectedVariable;
  return {
    ...state,
    dashboardState: { ...state.dashboardState, selectedVariable: newVarList }
  };
}

/**
 *
 *
 * @export
 * @param {{
 *   [key: string]: any;
 * }} data
 * @returns {ActionType}
 */
export function dashboardUpdateComputedResult(data: {
  variable: { [key: string]: any };
  recorder: { [key: string]: any };
  driver: { [key: string]: any };
}): ActionType {
  return {
    type: Action.DASHBOARD_UPDATE_COMPUTED_RESULT,
    data
  };
}

/**
 *
 *
 * @export
 * @param {StateInterface} state
 * @param {DashboardUpdateComputedResult} action
 * @returns {StateInterface}
 */
export function dashboardUpdateComputedResult_(
  state: StateInterface,
  action: DashboardUpdateComputedResult
): StateInterface {
  const computedResult = action.data.variable;
  const recorderData = action.data.recorder;
  const driverData = action.data.driver;
  return {
    ...state,
    dashboardState: {
      ...state.dashboardState,
      computedResult,
      recorderData,
      driverData
    }
  };
}
