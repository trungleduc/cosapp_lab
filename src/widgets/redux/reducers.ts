import { Action, ActionType, StateInterface } from './types';
import * as ActionFunc from './actions';

export const initialState: StateInterface = {
  systemConfig: {},
  mainState: {
    panel_0: [false, 'Project_Overview'],
    panel_1: [true, 'System_Arc'],
    panel_2: [false, 'Module_Creator'],
    panel_3: [false, 'Dashboard'],
    panel_4: [false, 'User_Management'],
    panel_5: [false, 'Module_Library'],
  },
  saveSignal: 0,
  systemArch: {
    lockStatus: false,
    systemTree: {
      nodeData: [{ title: 'Root' }],
      nodePath: [],
      selectedNode: null,
    },
    systemData: { mainData: {}, portData: {} },
    systemGraph: {
      systemGraphData: {
        Root: {
          inPort: ['inwards'],
          outPort: ['outwards'],
          connections: [],
        },
      },
      systemList: ['Root'],
      graphJsonData: {},
      updateData: {},
    },
    systemPBS: {},
    systemPBSUpdated: 0,
  },
  dashboardState: {
    computing: false,
    timeStepList: ['None'],
    data3D: {},
    variableData: {},
    portMetaData: {},
    selectedVariable: {},
    computedResult: {},
    recorderData: {},
    driverData : {}
  },
};

export function rootReducer(
  state: StateInterface = initialState,
  action: ActionType
): StateInterface {
  switch (action.type) {
    case Action.RESET_STORE: {
      return initialState;
    }

    case Action.SAVE_STATE: {
      return ActionFunc.saveState_(state);
    }

    case Action.SWITCH_PANEL: {
      return ActionFunc.mainSwitchPanel_(state, action);
    }

    case Action.LOCK_GRAPH: {
      return {
        ...state,
        systemArch: {
          ...state.systemArch,
          lockStatus: action.status,
        },
      };
    }

    case Action.GET_SERVER_DATA: {
      return ActionFunc.serverGetData_(state, action);
    }

    case Action.ARCH_SAVE_GRAPH_JSON: {
      return ActionFunc.archSaveGraphJson_(state, action);
    }

    case Action.ARCH_SAVE_GRAPH_POSITION: {
      return ActionFunc.archSaveGraphPosition_(state, action);
    }

    case Action.ARCH_UPDATE_CONNECTION_GRAPH: {
      return ActionFunc.archUpdateConnectionGraph_(state, action);
    }

    case Action.ARCH_FILTER_NODE: {
      return ActionFunc.archFilterNode_(state, action);
    }

    case Action.SELECT_NODE: {
      return ActionFunc.archSelectNode_(state, action);
    }

    case Action.ADD_NODE: {
      return ActionFunc.archAddNode_(state, action);
    }

    case Action.UPDATE_TREE: {
      return ActionFunc.archUpdateTreeData_(state, action);
    }

    case Action.REMOVE_NODE: {
      return ActionFunc.archRemoveNode_(state, action);
    }

    case Action.DASHBOARD_ADD_3D_DATA: {
      return ActionFunc.dashboardAdd3DData_(state, action);
    }

    case Action.DASHBOARD_TOGGLE_RUN: {
      return {
        ...state,
        dashboardState: {
          ...state.dashboardState,
          computing: !state.dashboardState.computing,
        },
      };
    }

    case Action.DASHBOARD_SET_CONTROLLER: {
      return ActionFunc.dashboardSetController_(state, action);
    }
    case Action.DASHBOARD_ADD_CONTROLLER: {
      return ActionFunc.dashboardAddController_(state, action);
    }

    case Action.DASHBOARD_REMOVE_CONTROLLER: {
      return ActionFunc.dashboardRemoveController_(state, action);
    }

    case Action.DASHBOARD_UPDATE_CONTROLLER: {
      return ActionFunc.dashboardUpdateController_(state, action);
    }

    case Action.DASHBOARD_UPDATE_COMPUTED_RESULT: {
      return ActionFunc.dashboardUpdateComputedResult_(state, action);
    }

  }

  return state;
}
