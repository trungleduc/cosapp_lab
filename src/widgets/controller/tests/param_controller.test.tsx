import { Provider } from 'react-redux';
import { initialState } from '../../redux/reducers';
import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import ParameterController from '../controller_element/param_controller';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
configure({ adapter: new Adapter() });

const mockStore = configureMockStore([thunk]);

describe('Test <ParameterController/>', () => {
  let wrapper: any;
  beforeEach(() => {
    initialState.dashboardState.portMetaData = {
      sys: {
        port: {
          variable: { desc: null, limits: null, valid_range: null, unit: null }
        }
      }
    };
    const store = mockStore({
      ...initialState
    });
    const varName = 'sys.port.variable';
    wrapper = mount(
      <Provider store={store}>
        <ParameterController variableName={varName} key={varName} value={0} />
      </Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('Should render correctly component', () => {
    const component = wrapper.find(ParameterController);
    expect(component).toHaveLength(1);
  });

  it('Should match snapshot', () => {
    const component = wrapper.find(ParameterController);
    expect(component.html()).toMatchSnapshot();
  });
});
