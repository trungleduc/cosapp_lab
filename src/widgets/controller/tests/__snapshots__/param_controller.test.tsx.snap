// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <ParameterController/> Should match snapshot 1`] = `"<div style=\\"padding: 10px 10px 10px 10px; border-bottom: 0.5px solid black;\\"><div></div><div></div><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center\\"><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7\\"><p class=\\"MuiTypography-root MuiTypography-body1\\" aria-haspopup=\\"true\\">sys.port.variable</p></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3\\"><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-marginDense MuiInput-marginDense\\"><input type=\\"number\\" aria-labelledby=\\"input-slider\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense\\" value=\\"0\\"></div></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2\\"><button class=\\"MuiButtonBase-root MuiIconButton-root\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"delete\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeSmall\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></div>"`;
